### PR TITLE
Adding support for latest OLMo architectures

### PR DESCRIPTION
**Pull request by natolambert**

### CODE REVIEW

File: Dockerfile.olmo [hunk 1]
```diff
@@ -0,0 +1,115 @@
+ARG CUDA
```
Comment (hamishivi) [inline Dockerfile.olmo:1]: do we want to add documentation about building with this image or nah? We had it before.

File: configs/train_configs/sft/olmo_7b_0924.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+model_name_or_path: ai2-adapt-dev/OLMo-medium-peteish7-anneal-from-928646-50B-nowup-dclm07-flan
+model_revision: main
+use_flash_attn: false
+tokenizer_name: ai2-adapt-dev/OLMo-medium-peteish7-anneal-from-928646-50B-nowup-dclm07-flan
```
Comment (hamishivi) [inline configs/train_configs/sft/olmo_7b_0924.yaml:4]: I would double check with luca that this is the right tokenizer. Sometimes conversions from the olmo repo and tokenizers get jumbled...

File: open_instruct/dpo_tune.py [hunk 3]
```diff
@@ -672,7 +679,7 @@ def load_model():
             0,
             1,
         ], "LlamaTokenizer should only add one special token - the pad_token, or no tokens if pad token present."
-    elif isinstance(tokenizer, GPTNeoXTokenizerFast):
+    elif isinstance(tokenizer, GPTNeoXTokenizerFast) or isinstance(tokenizer, OLMoTokenizerFast):
```
Comment (hamishivi) [inline open_instruct/dpo_tune.py]: wont this error if olmo is not imported? Maybe we should do:
`elif isinstance(tokenizer, GPTNeoXTokenizerFast) or (check_hf_olmo_availability() and isinstance(tokenizer, OLMoTokenizerFast)):`

File: open_instruct/finetune.py [hunk 4]
```diff
@@ -643,7 +650,7 @@ def main(args: FlatArguments):
             0,
             1,
         ], "LlamaTokenizer should only add one special token - the pad_token, or no tokens if pad token present."
-    elif isinstance(tokenizer, GPTNeoXTokenizerFast):
+    elif isinstance(tokenizer, GPTNeoXTokenizerFast) or isinstance(tokenizer, OLMoTokenizerFast): # noqa
```
Comment (hamishivi) [inline open_instruct/finetune.py]: same comment as above. `elif isinstance(tokenizer, GPTNeoXTokenizerFast) or (check_hf_olmo_availability() and isinstance(tokenizer, OLMoTokenizerFast)):`

File: open_instruct/utils.py [hunk 5]
```diff
@@ -51,6 +52,41 @@
...
+    pkg_name = "hf_olmo"
+
+    # Check if the package spec exists
+    package_exists = importlib.util.find_spec(pkg_name) is not None
+    package_version = "N/A"
+
+    if package_exists:
+        try:
+            # Primary method to get the package version
+            package_version = importlib.metadata.version(pkg_name)
```
Comment (hamishivi) [inline open_instruct/utils.py:69]: curious why doing this over like:
```python
try
    import hf_olmo
 except ImportError
    ...
```
is preferable. Don't mind it, just curious.

**Review by hamishivi — COMMENTED** [review]
Some comments... seems okay, but the tokenizer checks should be changed to not assume the user has hf_olmo installed.

**Review by vwxyzjn — COMMENTED** [review]
Looks good! Do you have a testing run?

**Comment by natolambert** [comment]
Yup @vwxyzjn have trained a couple models. Most recent one https://wandb.ai/ai2-llm/open_instruct_internal/runs/ny1m6zwx?nw=nwusernatolambert (plus another that already had eval scores, see slack)

**Review by vwxyzjn — APPROVED** [review]
LGTM! Nice change and thanks @natolambert!

**Comment by vwxyzjn** [comment]
Let's close it since we are training olmo2 with transformer's implementation.

**Pull request closed without merging**